src/main/java/org/apache/maven/plugins/toolchain/jdk/ToolchainDiscoverer.java

```diff
@@ -0,0 +1,468 @@
+        Xpp3Dom config = new Xpp3Dom("configuration");
+        Xpp3Dom jdkHome = new Xpp3Dom(JDK_HOME);
+        jdkHome.setValue(currentJdkHome.toString());
+        config.addChild(jdkHome);
+        model.setConfiguration(config);
+        return Optional.of(model);
+    }
+
+    public PersistedToolchains discoverToolchains() {
+        return discoverToolchains(LTS + COMMA + VERSION + COMMA + VENDOR);
+    }
+
+    /**
+     * Returns a PersistedToolchains object containing a list of discovered toolchains,
+     * never <code>null</code>.
+     */
+    public PersistedToolchains discoverToolchains(String comparator) {
+        try {
+            Set<Path> jdks = findJdks();
+            log.info("Found " + jdks.size() + " possible jdks: " + jdks);
+            readCache();
+            Map<Path, Map<String, String>> flags = new HashMap<>();
+            Path currentJdkHome = getCanonicalPath(Paths.get(System.getProperty(JAVA_HOME)));
+            flags.computeIfAbsent(currentJdkHome, p -> new HashMap<>()).put(CURRENT, "true");
+            // check environment variables for JAVA{xx}_HOME
+            System.getenv().entrySet().stream()
+                    .filter(e -> e.getKey().startsWith("JAVA") && e.getKey().endsWith("_HOME"))
+                    .forEach(e -> {
+                        Path path = getCanonicalPath(Paths.get(e.getValue()));
+                        Map<String, String> f = flags.computeIfAbsent(path, p -> new HashMap<>());
+                        String val = f.getOrDefault(ENV, "");
+                        f.put(ENV, (val.isEmpty() ? "" : val + ",") + e.getKey());
+                    });
+
+            List<ToolchainModel> tcs = jdks.parallelStream()
+                    .map(s -> {
+                        ToolchainModel tc = getToolchainModel(s);
+                        flags.getOrDefault(s, Collections.emptyMap())
+                                .forEach((k, v) -> tc.getProvides().setProperty(k, v));
+                        String version = tc.getProvides().getProperty(VERSION);
+                        if (isLts(version)) {
+                            tc.getProvides().setProperty(LTS, "true");
+                        }
+                        return tc;
+                    })
+                    .sorted(getToolchainModelComparator(comparator))
+                    .collect(Collectors.toList());
+            writeCache();
+            PersistedToolchains ps = new PersistedToolchains();
+            ps.setToolchains(tcs);
+            return ps;
+        } catch (Exception e) {
+            if (log.isDebugEnabled()) {
+                log.warn("Error discovering toolchains: " + e, e);
+            } else {
+                log.warn("Error discovering toolchains (enable debug level for more information): " + e);
+            }
+            return new PersistedToolchains();
+        }
+    }
+
+    private static boolean isLts(String version) {
+        return version.startsWith("1.8.")
```

Comment:
sorry in advance if silly question.
will that work with sdkman for 1.8/8?
Because sdkman is a bit strange for 1.8 jdk.
Local install is `/home/olamy/.sdkman/candidates/java/8.0.372-tem`
But obviously: 
```
 /home/olamy/.sdkman/candidates/java/8.0.372-tem/bin/java -version
openjdk version "1.8.0_372"
OpenJDK Runtime Environment (Temurin)(build 1.8.0_372-b07)
OpenJDK 64-Bit Server VM (Temurin)(build 25.372-b07, mixed mode)
```

could it be a problem? 1.8 vs 8?

Reply:
I don't have this particular version available on sdkman, however:
```
➜  maven-toolchains-plugin git:(discovered-toolchains) mvn org.apache.maven.plugins:maven-toolchains-plugin:3.1.1-SNAPSHOT:display-discovered-jdk-toolchains 
Using MAVEN_CENTRAL=http://old-broken-macbook-pro.local:8082/artifactory/maven-remote/
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------< org.apache.maven.plugins:maven-toolchains-plugin >----------
[INFO] Building Apache Maven Toolchains Plugin 3.1.1-SNAPSHOT
[INFO]   from pom.xml
[INFO] ----------------------------[ maven-plugin ]----------------------------
[INFO] 
[INFO] --- toolchains:3.1.1-SNAPSHOT:display-discovered-jdk-toolchains (default-cli) @ maven-toolchains-plugin ---
[INFO] Found 11 possible jdks: [/Users/gnodet/.sdkman/candidates/java/21.0.2-tem, /Users/gnodet/.jabba/jdk/zulu@1.8.282/Contents/Home, /Users/gnodet/.sdkman/candidates/java/8.0.402-librca, /Users/gnodet/.jbang/cache/jdks/17, /Users/gnodet/.sdkman/candidates/java/21.0.2-graalce, /Users/gnodet/.sdkman/candidates/java/17.0.9-graalce, /Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home, /Users/gnodet/.sdkman/candidates/java/11.0.22-tem, /Users/gnodet/.sdkman/candidates/java/17.0.10-tem, /Users/gnodet/.sdkman/candidates/java/22.ea.36-open, /Users/gnodet/.sdkman/candidates/java/23.ea.13-open]
[INFO] Discovered 11 JDK toolchains:
[INFO]   - /Users/gnodet/.sdkman/candidates/java/21.0.2-graalce
[INFO]     provides:
[INFO]       version: 21.0.2
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 21.0.2+13-jvmci-23.1-b30
[INFO]       vendor: GraalVM Community
[INFO]       vendor.version: GraalVM CE 21.0.2+13.1
[INFO]       current: true
[INFO]       lts: true
[INFO]       env: JAVA_HOME
[INFO]   - /Users/gnodet/.sdkman/candidates/java/21.0.2-tem
[INFO]     provides:
[INFO]       version: 21.0.2
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 21.0.2+13-LTS
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-21.0.2+13
[INFO]       lts: true
[INFO]   - /Users/gnodet/.sdkman/candidates/java/17.0.9-graalce
[INFO]     provides:
[INFO]       version: 17.0.9
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 17.0.9+9-jvmci-23.0-b22
[INFO]       vendor: GraalVM Community
[INFO]       vendor.version: GraalVM CE 17.0.9+9.1
[INFO]       lts: true
[INFO]   - /Users/gnodet/.jbang/cache/jdks/17
[INFO]     provides:
[INFO]       version: 17.0.7
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 17.0.7+7
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-17.0.7+7
[INFO]       lts: true
[INFO]   - /Users/gnodet/.sdkman/candidates/java/17.0.10-tem
[INFO]     provides:
[INFO]       version: 17.0.10
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 17.0.10+7
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-17.0.10+7
[INFO]       lts: true
[INFO]   - /Users/gnodet/.sdkman/candidates/java/11.0.22-tem
[INFO]     provides:
[INFO]       version: 11.0.22
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 11.0.22+7
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-11.0.22+7
[INFO]       lts: true
[INFO]   - /Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home
[INFO]     provides:
[INFO]       version: 1.8.0_402
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 1.8.0_402-b06
[INFO]       vendor: Azul Systems, Inc.
[INFO]       lts: true
[INFO]   - /Users/gnodet/.sdkman/candidates/java/8.0.402-librca
[INFO]     provides:
[INFO]       version: 1.8.0_402
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 1.8.0_402-b07
[INFO]       vendor: BellSoft
[INFO]       lts: true
[INFO]   - /Users/gnodet/.jabba/jdk/zulu@1.8.282/Contents/Home
[INFO]     provides:
[INFO]       version: 1.8.0_282
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 1.8.0_282-b08
[INFO]       vendor: Azul Systems, Inc.
[INFO]       lts: true
[INFO]   - /Users/gnodet/.sdkman/candidates/java/23.ea.13-open
[INFO]     provides:
[INFO]       version: 23-ea
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 23-ea+13-981
[INFO]       vendor: Oracle Corporation
[INFO]   - /Users/gnodet/.sdkman/candidates/java/22.ea.36-open
[INFO]     provides:
[INFO]       version: 22
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 22+36-2370
[INFO]       vendor: Oracle Corporation
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.167 s
[INFO] Finished at: 2024-03-19T18:35:25+01:00
[INFO] ------------------------------------------------------------------------
```

The best thing is, if you have two minutes, build this PR and run `mvn org.apache.maven.plugins:maven-toolchains-plugin:3.1.1-SNAPSHOT:display-discovered-jdk-toolchains` to check if it's correctly discovered.

Reply:
all good!!
```
olamy@pop-os:~/dev/sources/maven/plugins/tools/maven-toolchains-plugin$ mvn org.apache.maven.plugins:maven-toolchains-plugin:3.1.1-SNAPSHOT:display-discovered-jdk-toolchains
[INFO] Scanning for projects...
[WARNING] The requested profile "outside" could not be activated because it does not exist.
[INFO] 
[INFO] ----------< org.apache.maven.plugins:maven-toolchains-plugin >----------
[INFO] Building Apache Maven Toolchains Plugin 3.1.1-SNAPSHOT
[INFO]   from pom.xml
[INFO] ----------------------------[ maven-plugin ]----------------------------
[INFO] 
[INFO] --- toolchains:3.1.1-SNAPSHOT:display-discovered-jdk-toolchains (default-cli) @ maven-toolchains-plugin ---
[INFO] Found 13 possible jdks: [/home/olamy/.sdkman/candidates/java/21.0.1-tem, /home/olamy/.sdkman/candidates/java/11.0.21-tem, /home/olamy/.sdkman/candidates/java/17.0.10-tem, /home/olamy/.sdkman/candidates/java/22.ea.36-open, /home/olamy/.sdkman/candidates/java/21-open, /home/olamy/.sdkman/candidates/java/21.0.2-tem, /home/olamy/.sdkman/candidates/java/17.0.9-tem, /home/olamy/.sdkman/candidates/java/21-tem, /home/olamy/.sdkman/candidates/java/11.0.22-tem, /home/olamy/.sdkman/candidates/java/8.0.372-tem, /home/olamy/softs/jdks/zulu17.44.17-ca-crac-jdk17.0.8-linux_x64, /home/olamy/.sdkman/candidates/java/22.ea.34-open, /home/olamy/.sdkman/candidates/java/11.0.19-tem]
[INFO] Discovered 13 JDK toolchains:
[INFO]   - /home/olamy/.sdkman/candidates/java/11.0.22-tem
[INFO]     provides:
[INFO]       version: 11.0.22
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 11.0.22+7
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-11.0.22+7
[INFO]       current: true
[INFO]       lts: true
[INFO]       env: JAVA_HOME
[INFO]   - /home/olamy/.sdkman/candidates/java/21.0.2-tem
[INFO]     provides:
[INFO]       version: 21.0.2
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 21.0.2+13-LTS
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-21.0.2+13
[INFO]       lts: true
[INFO]   - /home/olamy/.sdkman/candidates/java/21.0.1-tem
[INFO]     provides:
[INFO]       version: 21.0.1
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 21.0.1+12-LTS
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-21.0.1+12
[INFO]       lts: true
[INFO]   - /home/olamy/.sdkman/candidates/java/17.0.9-tem
[INFO]     provides:
[INFO]       version: 17.0.9
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 17.0.9+9
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-17.0.9+9
[INFO]       lts: true
[INFO]   - /home/olamy/softs/jdks/zulu17.44.17-ca-crac-jdk17.0.8-linux_x64
[INFO]     provides:
[INFO]       version: 17.0.8
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 17.0.8+7-LTS
[INFO]       vendor: Azul Systems, Inc.
[INFO]       vendor.version: Zulu17.44+17-CRaC-CA
[INFO]       lts: true
[INFO]   - /home/olamy/.sdkman/candidates/java/17.0.10-tem
[INFO]     provides:
[INFO]       version: 17.0.10
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 17.0.10+7
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-17.0.10+7
[INFO]       lts: true
[INFO]   - /home/olamy/.sdkman/candidates/java/11.0.21-tem
[INFO]     provides:
[INFO]       version: 11.0.21
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 11.0.21+9
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-11.0.21+9
[INFO]       lts: true
[INFO]   - /home/olamy/.sdkman/candidates/java/11.0.19-tem
[INFO]     provides:
[INFO]       version: 11.0.19
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 11.0.19+7
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-11.0.19+7
[INFO]       lts: true
[INFO]   - /home/olamy/.sdkman/candidates/java/8.0.372-tem
[INFO]     provides:
[INFO]       version: 1.8.0_372
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 1.8.0_372-b07
[INFO]       vendor: Temurin
[INFO]       lts: true
[INFO]   - /home/olamy/.sdkman/candidates/java/22.ea.34-open
[INFO]     provides:
[INFO]       version: 22-ea
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 22-ea+34-2360
[INFO]       vendor: Oracle Corporation
[INFO]   - /home/olamy/.sdkman/candidates/java/22.ea.36-open
[INFO]     provides:
[INFO]       version: 22
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 22+36-2370
[INFO]       vendor: Oracle Corporation
[INFO]   - /home/olamy/.sdkman/candidates/java/21-tem
[INFO]     provides:
[INFO]       version: 21
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 21+35-LTS
[INFO]       vendor: Eclipse Adoptium
[INFO]       vendor.version: Temurin-21+35
[INFO]   - /home/olamy/.sdkman/candidates/java/21-open
[INFO]     provides:
[INFO]       version: 21
[INFO]       runtime.name: OpenJDK Runtime Environment
[INFO]       runtime.version: 21+35-2513
[INFO]       vendor: Oracle Corporation

```
Thanks

Reply:
> sorry in advance if silly question. will that work with sdkman for 1.8/8? Because sdkman is a bit strange for 1.8 jdk. Local install is `/home/olamy/.sdkman/candidates/java/8.0.372-tem` But obviously:
> 
> ```
>  /home/olamy/.sdkman/candidates/java/8.0.372-tem/bin/java -version
> openjdk version "1.8.0_372"
> OpenJDK Runtime Environment (Temurin)(build 1.8.0_372-b07)
> OpenJDK 64-Bit Server VM (Temurin)(build 25.372-b07, mixed mode)
> ```
> 
> could it be a problem? 1.8 vs 8?

Mmh, I misread your comment.  The `isLts()` check is done against the `version` property _from the toolchain_, but you're right, it will fail against versions such as `22`.  I'll make it more robust.